statistics/handle/cache/internal/lfu/lfu_cache_test.go

```diff
@@ -0,0 +1,106 @@
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	mockCMSMemoryUsage  = int64(4)
+	mockTopNMemoryUsage = int64(64)
+	mockHistMemoryUsage = int64(289)
+)
+
+func TestLFUPutGetDel(t *testing.T) {
+	capacity := int64(100)
+	lru, err := NewLFU(capacity)
+	require.NoError(t, err)
+	mockTable := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	mockTableID := int64(1)
+	lru.Put(mockTableID, mockTable, false)
+	require.Eventually(t, func() bool {
+		v, ok := lru.Get(mockTableID, false)
+		return ok && mockTable == v
+	}, 1*time.Second, 100*time.Millisecond)
+
+	lru.Del(mockTableID)
+	v, ok := lru.Get(mockTableID, false)
+	require.False(t, ok)
+	require.Nil(t, v)
+}
+
+func TestLFUFreshMemUsage(t *testing.T) {
+	lru, err := NewLFU(1000)
+	require.NoError(t, err)
+	t1 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	t2 := testutil.NewMockStatisticsTable(2, 2, true, false, false)
+	t3 := testutil.NewMockStatisticsTable(3, 3, true, false, false)
+	lru.Put(int64(1), t1, false)
+	lru.Put(int64(2), t2, false)
+	lru.Put(int64(3), t3, false)
+	time.Sleep(1 * time.Second)
+	require.Equal(t, lru.Cost(), 6*mockCMSMemoryUsage+6*mockCMSMemoryUsage)
+	t4 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	lru.Put(int64(1), t4, false)
+	time.Sleep(4 * time.Second)
+	require.Equal(t, lru.Cost(), 6*mockCMSMemoryUsage+7*mockCMSMemoryUsage)
+	t5 := testutil.NewMockStatisticsTable(2, 2, true, false, false)
+	lru.Put(int64(1), t5, false)
+	require.Equal(t, lru.Cost(), 7*mockCMSMemoryUsage+7*mockCMSMemoryUsage)
+
+	t6 := testutil.NewMockStatisticsTable(1, 2, true, false, false)
+	lru.Put(int64(1), t6, false)
+	require.Equal(t, lru.Cost(), 7*mockCMSMemoryUsage+6*mockCMSMemoryUsage)
+
+	t7 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	lru.Put(int64(1), t7, false)
+	require.Equal(t, lru.Cost(), 6*mockCMSMemoryUsage+6*mockCMSMemoryUsage)
+}
+
+func TestLFUPutTooBig(t *testing.T) {
+	lru, err := NewLFU(1)
+	require.NoError(t, err)
+	mockTable := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	// put mockTable, the index should be evicted
+	lru.Put(int64(1), mockTable, false)
+	_, ok := lru.Get(int64(1), false)
+	require.False(t, ok)
+}
+
+func TestCacheLen(t *testing.T) {
+	capacity := int64(12)
+	lru, err := NewLFU(capacity)
+	require.NoError(t, err)
+	t1 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	require.Equal(t, int64(12), t1.MemoryUsage().TotalTrackingMemUsage())
+	lru.Put(int64(1), t1, false)
+	time.Sleep(1 * time.Second)
+	t2 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	// put t2, t1 should be evicted 2 items and still exists in the list
+	lru.Put(int64(2), t2, false)
+	time.Sleep(5 * time.Second)
+	require.Equal(t, lru.Len(), 1)
+
+	// put t3, t1/t2 should be evicted all items and disappeared from the list
+	t3 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	lru.Put(int64(3), t3, false)
+	time.Sleep(1 * time.Second)
+	require.Equal(t, lru.Len(), 1)
+}
```

Comment:
Better to add some concurrent test cases, e.g. start some goroutines to read/write this cache and see whether there is any DATA RACE and whether the status is still consistent after some operations.

Reply:
Done